src/videojs-hls.js

```diff
@@ -465,15 +507,29 @@ videojs.Hls.prototype.updateDuration = function(playlist) {
         this.mediaSource.duration = newDuration;
         this.tech_.trigger('durationchange');
         this.mediaSource.removeEventListener('sourceopen', setDuration);
-      }.bind(this);
+        if (this.sourceBuffer) {
+          this.sourceBuffer.removeEventListener('updatend', setDuration);
+        }
+      }.bind(this),
+      seekable = this.seekable();
+
+  if (seekable.length && newDuration === Infinity) {
+    if (isNaN(oldDuration)) {
+      oldDuration = 0;
+    }
+    newDuration = Math.max(
+      oldDuration, seekable.end(0) + playlist.targetDuration * 3);
+  }
 
   // if the duration has changed, invalidate the cached value
   if (oldDuration !== newDuration) {
-    if (this.mediaSource.readyState === 'open') {
+    if (this.mediaSource.readyState !== 'open') {
+      this.mediaSource.addEventListener('sourceopen', setDuration);
+    } else if (this.sourceBuffer && this.sourceBuffer.updating) {
+      this.sourceBuffer.addEventListener('updateend', setDuration);
```

Comment:
There may be multiple `updateend` listeners on a source buffer and one of the previous ones could perform an operation that causes it to enter the updating state.